lib/auth/state_unix.go

```diff
@@ -38,9 +43,76 @@ func NewProcessStorage(ctx context.Context, path string) (*ProcessStorage, error
 		EventsOff: true,
 		Sync:      lite.SyncFull,
 	})
+
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
 
-	return &ProcessStorage{Backend: litebk}, nil
+	// if running in a K8S cluster and required env vars are available
+	// the agent will automatically switch state storage from local
+	// sqlite into a Kubernetes Secret.
+	if kubernetes.InKubeCluster() {
+		kubeStorage, err := kubernetes.New()
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		// if secret does not exist but Statefulset had local storage with identities,
+		// the agent reads them from SQLite and dumps into Kubernetes Secret.
+		// TODO(tigrato): remove this once the compatibility layer between local
+		// storage and Kube secret storage is no longer required!
+		if !kubeStorage.Exists(ctx) {
+			if err := copyLocalStorageIntoKubernetes(ctx, kubeStorage, litebk); err != nil {
+				return nil, trace.Wrap(err)
+			}
+		}
+
+		identityStorage = kubeStorage
+	} else {
+		// if not in Kubernetes or env vars are not available use local sqlite as identity backend
+		identityStorage = litebk
+	}
+
+	return &ProcessStorage{BackendStorage: litebk, stateStorage: identityStorage}, nil
+}
+
+// copyLocalStorageIntoKubernetes reads every `identity` and `state` keys from local storage
+// and copies them into Kubernetes Secret. This code is executed only when the agent starts and the
+// secret was not yet created in K8S. Subsequent restarts of the agent won't execute this code since the
+// secret already exists in K8S
+// TODO(tigrato): remove this once the compatibility layer between local storage and
+// Kube secret storage is no longer required!
+func copyLocalStorageIntoKubernetes(ctx context.Context, k8sStorage *kubernetes.Backend, litebk *lite.Backend) error {
```

Comment:
What happens if we get a failure during `PutItems`? Is there a way to have this heal itself in some way (by creating some key at the beginning and only deleting as the last step, for instance) instead of requiring manual intervention to delete the half-migrated secrets?

Reply:
Nice point. 
`PutItems` is only called if the secret does not exist and it will dump every key found in the SQLite in a single shot. 
I changed the secret creation to push directly the new keys so we no longer have empty secrets that might cause "half-migrated" secrets.